### PR TITLE
[expo-updates][ios] track build configuration data

### PR DESCRIPTION
**Pull request by jkhales**
# Why

 The build data stored by the configuration is subject to change when a user updates the binary.
 
 This can lead to inconsistent update loading behavior, for example if a user were to publish to channel A after a build with channel B was time stamped, then when the client downloaded build B they would still get update A, instead of the expected build B embedded update:
 
Screen Shot 2021-11-04 at 11 02 58 AM
 
 c.f. https://github.com/expo/expo/issues/14372
 
 
 
# How

Introduce a singleton that wipes the updates when any of the tracked build data changes. This leaves the user in the same situation as a fresh install.
 
While the original report is for `releaseChannel` and also applies to `requestHeaders.channel`, a few extra fields are included for safety after discussion with @esamelson 

# Test Plan
(as always with expo-updates, make sure you are building for release and not debug)

1. create build
2. publish update
3. load update
4.  change one of the fields tracked by `buildData` such as `releaseChannel`
5. edit the commitTime in `scripts/createManifest.js` back before the time of the initial build and publish.
6. rebuild  and load without deleting the previous

Since this is back in time, if the updates are not cleared, the published update will load. With this fix applied the embedded update from step 5 should load.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-updates/ios/EXUpdates/Database/EXUpdatesDatabase.m [hunk 1]
```diff
@@ -460,7 +461,7 @@ - (nullable NSDictionary *)_jsonDataWithKey:(NSString *)key scopeKey:(NSString *
     id value = rows[0][@"value"];
     if (value && [value isKindOfClass:[NSString class]]) {
       NSDictionary *jsonObject = [NSJSONSerialization JSONObjectWithData:[(NSString *)value dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:error];
-      if (!*error && jsonObject && [jsonObject isKindOfClass:[NSDictionary class]]) {
+      if (!(error && *error) && jsonObject && [jsonObject isKindOfClass:[NSDictionary class]]) {
```
Comment (jkhales) [inline packages/expo-updates/ios/EXUpdates/Database/EXUpdatesDatabase.m:464]: (pre-existing bug)

File: packages/expo-updates/ios/EXUpdates/EXUpdatesAppController.m [hunk 2]
```diff
@@ -144,8 +145,17 @@ - (void)start
     return;
   }
 
+  NSError *buildDataError;
+  [EXUpdatesBuildData ensureBuildDataIsConsistent:_database config:_config error:&buildDataError];
+  if (buildDataError) {
+    @throw [NSException exceptionWithName:NSInternalInconsistencyException
+                                   reason:buildDataError.localizedDescription
+                                 userInfo:buildDataError.userInfo];
```
Comment (jkhales) [inline packages/expo-updates/ios/EXUpdates/EXUpdatesAppController.m:153]: curious in general about your thoughts on how these errors are handled in this PR.

**Comment by jkhales** [comment]
Putting up just the iOS while I keep working on that weird stale context issue for android.

File: packages/expo-updates/ios/Tests/EXUpdatesBuildDataTests.m [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+    XCTAssertTrue([staticBuildData isEqualToDictionary:[EXUpdatesBuildData getBuildDataFromConfig:_configChannelTest]]);
+    NSArray<EXUpdatesUpdate *> *allUpdates = [_db allUpdatesWithConfig:_configChannelTest error:nil];
+    XCTAssertEqual(allUpdates.count, 1);
+    
+    XCTAssertNil(error);
+  });
+}
+
+- (void)test_ensureBuildDataIsConsistent_buildDataIsInconsistent {
```
Comment (esamelson) [inline packages/expo-updates/ios/Tests/EXUpdatesBuildDataTests.m:167]: do you think it's worth separately testing that we get the correct behavior if `releaseChannel` is inconsistent and there's no `channel` defined?

File: packages/expo-updates/ios/EXUpdates/Database/EXUpdatesBuildData.m [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+{
+  __block NSError *dbError;
+  dispatch_async(database.databaseQueue, ^{
+    NSArray<EXUpdatesUpdate *> *allUpdates = [database allUpdatesWithConfig:config error:&dbError];
+    if (allUpdates || !dbError) {
+      [database deleteUpdates:allUpdates error:&dbError];
+    }
+  });
+  if (dbError){
+    *error = dbError;
```
Comment (esamelson) [inline packages/expo-updates/ios/EXUpdates/Database/EXUpdatesBuildData.m:90]: same comment as above -- this won't work correctly with the dispatch_async. If this is a utility method only for tests, you could potentially just get rid of the dispatching entirely and require the caller to be responsible for being on the correct queue. (You could also consider just defining a `deleteAllUpdates` method on the database directly.)

File: packages/expo-updates/ios/Tests/EXUpdatesBuildDataTests.m [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+- (void)test_ensureBuildDataIsConsistent_buildDataIsNull {
+  // check no updates and build data is set
+
+  dispatch_sync(_db.databaseQueue, ^{
+    NSError *error;
+    NSDictionary *staticBuildData = [EXUpdatesBuildData getBuildDataFromDatabase:_db scopeKey:scopeKey error:&error];
+    XCTAssertNil(staticBuildData);
+
+    NSArray<EXUpdatesUpdate *> *allUpdates = [_db allUpdatesWithConfig:_configChannelTest error:&error];
+    XCTAssertEqual(allUpdates.count, 0);
```
Comment (esamelson) [inline packages/expo-updates/ios/Tests/EXUpdatesBuildDataTests.m:122]: do you think it's worth testing the case where there are updates (ensuring they don't get cleared)?

**Review by esamelson — APPROVED** [review]
This looks like a great start -- and thanks for adding tests! 🙏  My comments are mostly around the GCD stuff. Lmk if you want to discuss or if I can clarify anything more.

**Review by esamelson — CHANGES_REQUESTED** [review]
(oops, meant to request changes rather than approving 😅 )

File: packages/expo-updates/ios/Tests/EXUpdatesBuildDataTests.m (hunk 5 above)
Comment (jkhales) [reply packages/expo-updates/ios/Tests/EXUpdatesBuildDataTests.m:124]: yea good point, I ended up adding an update to each test and checking that it was either cleared or left alone as intended.

File: packages/expo-updates/ios/Tests/EXUpdatesBuildDataTests.m [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+    NSArray<EXUpdatesUpdate *> *allUpdates = [_db allUpdatesWithConfig:_configChannelTest error:nil];
+    XCTAssertEqual(allUpdates.count, 1);
+    XCTAssertNil(error);
+  });
+}
+ 
+- (void)test_ensureBuildDataIsConsistent_buildDataIsConsistent_releaseChannel {
+  dispatch_sync(_db.databaseQueue, ^{
+    NSError *error;
+    NSArray<EXUpdatesUpdate *> *allUpdates = [_db allUpdatesWithConfig:_configChannelTest error:&error];
```
Comment (esamelson) [inline packages/expo-updates/ios/Tests/EXUpdatesBuildDataTests.m:197]:
```suggestion
    NSArray<EXUpdatesUpdate *> *allUpdates = [_db allUpdatesWithConfig:_configReleaseChannelTest error:&error];
```

**Review by esamelson — APPROVED** [review]
Just a few minor comments & nits, but otherwise looks good!

**Pull request merged**